0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+that surrogate `char` values are only used in pairs as intended.
+This would likely require more user effort to migrate their code, however,
+as every operation on strings would now have to prove that a surrogate pair
+is not being split up. 
+It also does not reduce the complexity of compiling
+strings to various target languages that do not necessarily use UTF-16.
+
+## Define "valid UTF-16 strings" as a subset type in a shared library
+
+The next cheapest solution would be to define a subset type of `string` that enforces
+correct usage of surrogate characters.
+Similar such definitions already exist for valid [UTF-8](https://github.com/aws/aws-encryption-sdk-dafny/blob/mainline/src/Util/UTF8.dfy#L21)
+and [UTF-16](https://github.com/dafny-lang/libraries/blob/master/src/Unicode/Utf16EncodingForm.dfy) code unit sequences
+in some Dafny codebases.
+This also introduces the same proof obligations as the previous solution,
+although here they could be proved by numerous helper methods that operate on the subset type
+in the same shared library.
+
+This would likely be the best solution if changing the Dafny language itself was not an option.
+The major downside is that built-in sequence operations such as `s[i]` and `s + s'` would still require additional
+effort to achieve verification, or would have to be abandoned entirely in favour of the helper methods
+provided by the shared library.
+
+## Change the definition of the `string` type
+
+The current definition of `string` as an alias for `seq<char>` is problematic because surrogate code points
+are not truly independent values in the sequence.
+It is actually quite unusual for a general-purpose language to treat strings as only a special case of
+generic datatypes for arbitrary sequences of values.
+Dafny could instead make `string` an alias for a different type, whether built-in or defined in Dafny source in the runtime or a standard library,
+that treats strings as a more specialized datatype. It could then be more prescriptive about picking a definitive encoding to
+support directly (most likely UTF-8).
+
+This would certainly require more effort for existing code to migrate to.
+It is also a much more expensive solution to implement, since it requires creating an API for strings
+that is at least adequate to replace all existing usage of sequence operations.
+These interfaces often become very large to support all common string operations efficiently,
+and would need very careful thought.
+Although unusual, representing strings directly as sequences of abstract characters
+works very well for a verification-focussed language like Dafny,
+since sequences are already a deep built-in concept in the language semantics.
+
+This new type could alternatively be introduced with a different name, such as `unicode` as in Python 2,
+while keeping the alias of `string` for `seq<char>`.
+This would only increase the confusion and cognitive burden for Dafny users in the future, though.
```

Comment:
Can we discuss how easy it will be for external code to wrap a native string type to give it Dafny's string API?  It would be nice if one could implement `ISequence<char>` using C# strings (plus some sanity checks), to save on conversions.

Reply:
I'm going to rewrite the `Compilation/Runtime` section to be more clear about this (as I intended that to address exactly this point, but I don't think it's clear enough yet)

Reply:
I THINK this should have enough detail but let me know otherwise.